tls/s2n_x509_validator.c

```diff
@@ -616,6 +619,14 @@ S2N_RESULT s2n_x509_validator_validate_cert_chain(struct s2n_x509_validator *val
         RESULT_GUARD_POSIX(s2n_extension_list_process(S2N_EXTENSION_LIST_CERTIFICATE, conn, &first_certificate_extensions));
     }
 
+    if (conn->config->cert_validation_cb) {
+        struct s2n_cert_validation_info info = { 0 };
+        int ret = conn->config->cert_validation_cb(conn, &info, conn->config->cert_validation_ctx);
+        RESULT_ENSURE(ret == S2N_SUCCESS, S2N_ERR_CANCELLED);
+        RESULT_ENSURE(info.finished, S2N_ERR_INVALID_STATE);
+        RESULT_ENSURE(info.accepted, S2N_ERR_CERT_VALIDATION_CALLBACK_FAILED);
+    }
```

Comment:
Since the other callbacks can also error with `S2N_ERR_CANCELED`, I was thinking it might be good to have a way for applications to know for sure that their callback rejected the cert. Though, to provide any meaningful logging they would have to set this information in their connection context anyway, so maybe a separate error isn't that useful?

Reply:
That sounds reasonable, but I think the name / message is deceptive. https://github.com/aws/s2n-tls/pull/4156#discussion_r1300730710